### PR TITLE
Intersection improvements

### PR DESCRIPTION
**Pull request by smoogipooo**
Might as well get these performance improvements in before I do more work on occlusion testing, because the performance difference is measurable.

Old (left) vs new (right) (one scroll-through the visual test):
image

### CODE REVIEW

File: osu.Framework/Graphics/Primitives/Quad.cs [hunk 1]
```diff
@@ -16,21 +15,50 @@ public struct Quad : IConvexPolygon, IEquatable<Quad>
...
+        public int AxisCount { get; }
+
         public Quad(Vector2 topLeft, Vector2 topRight, Vector2 bottomLeft, Vector2 bottomRight)
         {
             TopLeft = topLeft;
             TopRight = topRight;
             BottomLeft = bottomLeft;
             BottomRight = bottomRight;
+
+            LeftAxis = new Axis(bottomLeft, topLeft);
```
Comment (Tom94) [inline osu.Framework/Graphics/Primitives/Quad.cs]: Doesn't this add quite some overhead everywhere we use Quads? (Which is in a lot of places.) Also the memory footprint goes up quite a bit from this (tripled).

File: osu.Framework/Graphics/Primitives/Triangle.cs [hunk 2]
```diff
@@ -6,17 +6,39 @@
...
+
         public Triangle(Vector2 p0, Vector2 p1, Vector2 p2)
         {
             P0 = p0;
             P1 = p1;
             P2 = p2;
+
+            Axis01 = new Axis(P0, P1);
+            Axis12 = new Axis(P1, P2);
+            Axis20 = new Axis(P2, P0);
```
Comment (Tom94) [inline osu.Framework/Graphics/Primitives/Triangle.cs]: Same here. I feel like we should just compute these guys locally whenever we want to perform intersections.

Comment (Tom94) [inline osu.Framework/Graphics/Primitives/Triangle.cs]: Oh wait, C# doesn't support dynamically sized data on the stack. x_x

Comment (Tom94) [inline osu.Framework/Graphics/Primitives/Triangle.cs]: In fact, you should compute `Axis` locally inside `GetVertex`. That should be next to zero computational overhead by avoiding heap allocs. Also, that avoids the problem of having out-of-date axes as soon as someone changes vertices of the triangle. Same with quads.

**Comment by peppy** [comment]
shall we close this for now?

**Comment by smoogipooo** [comment]
Sure

**Pull request closed without merging**